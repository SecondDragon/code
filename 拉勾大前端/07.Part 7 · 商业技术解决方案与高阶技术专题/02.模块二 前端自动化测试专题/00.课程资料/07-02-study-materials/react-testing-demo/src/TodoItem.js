import React, { useState, useRef } from 'react'
import classNames from 'classnames'

export default function TodoItem ({ todo, doneTodo, removeTodo, saveTodo }) {
  const [isEdit, setIsEdit] = useState(false)
  const inputEl = useRef(null)

  const saveEdit = e => {
    const value = e.target.value.trim()
    if (value.length) {
      saveTodo(todo.id, e.target.value)
    } else {
      removeTodo(todo.id)
    }
    setIsEdit(false)
  }

  const cancelEdit = e => {
    setIsEdit(false)
    e.target.value = todo.text
  }

  return (
    <li className={classNames({
      completed: todo.done,
      editing: isEdit
    })} data-testid="todo-item">
      {/* These are here just to show the structure of the list items */}
      {/* List items should get the class `editing` when editing and `completed` when marked as completed */}
      <div className="view">
        <input
          className="toggle"
          type="checkbox"
          checked={todo.done}
          data-testid="todo-done"
          onChange={e => doneTodo(todo.id, e.target.checked)}
        />
        <label
          data-testid="todo-text"
          onDoubleClick={() => {
            setIsEdit(true)
            setTimeout(() => {
              inputEl.current.focus()
            }, 0)
          }}
        >{todo.text}</label>
        <button
          className="destroy"
          data-testid="destroy"
          onClick={() => removeTodo(todo.id)}
        ></button>
      </div>
      <input
        ref={inputEl}
        className="edit"
        autoFocus
        defaultValue={todo.text}
        onBlur={saveEdit}
        data-testid="todo-edit"
        onKeyUp={e => {
          if (e.key === 'Enter') { // 保存编辑
            saveEdit(e)
          } else if (e.key === 'Escape') { // 取消编辑
            cancelEdit(e)
          }
        }}
      />
    </li>
  )
}
