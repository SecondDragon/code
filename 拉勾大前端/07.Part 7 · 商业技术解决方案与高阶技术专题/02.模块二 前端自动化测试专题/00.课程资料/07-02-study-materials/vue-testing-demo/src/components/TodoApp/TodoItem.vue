<template>
  <li data-testid="todo-item" :class="{
    completed: todo.done,
    editing: isEditing
  }">
    <div class="view">
      <input v-model="todo.done" data-testid="todo-done" class="toggle" type="checkbox" />
      <label
        data-testid="todo-text"
        @dblclick="isEditing = true"
      >{{ todo.text }}</label>
      <button
        data-testid="delete"
        class="destroy"
        @click="$emit('delete-todo', todo.id)"
      ></button>
    </div>
    <input
      :value="todo.text"
      v-focus="isEditing"
      class="edit"
      data-testid="todo-edit"
      @blur="isEditing = false"
      @keyup.enter="handleEditTodo"
      @keyup.esc="handleCancelEdit"
    />
  </li>
</template>

<script>
export default {
  name: 'TodoItem',
  props: {
    todo: {
      type: Object,
      required: true
    }
  },
  directives: {
    focus (element, binding) {
      if (binding.value) {
        element.focus()
      }
    }
  },
  data () {
    return {
      isEditing: false
    }
  },
  methods: {
    handleEditTodo (e) {
      this.$emit('edit-todo', {
        id: this.todo.id,
        text: e.target.value
      })
      // 取消编辑状态
      this.isEditing = false
    },

    handleCancelEdit () {
      this.isEditing = false
    }
  }
}
</script>
