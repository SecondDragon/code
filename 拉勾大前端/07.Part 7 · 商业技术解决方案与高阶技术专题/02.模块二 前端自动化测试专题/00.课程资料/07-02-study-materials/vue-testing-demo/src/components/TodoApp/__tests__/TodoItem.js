import { shallowMount } from '@vue/test-utils'
import TodoItem from '@/components/TodoApp/TodoItem.vue'

describe('TodoItem.vue', () => {
  /** @type {import('@vue/test-utils').Wrapper} */
  let wrapper = null
  beforeEach(() => {
    const todo = {
      id: 1,
      text: 'play',
      done: true
    }
    wrapper = shallowMount(TodoItem, {
      propsData: {
        todo
      }
    })
  })
  test('text', () => {
    expect(wrapper.find('[data-testid="todo-text"]').text()).toBe(wrapper.vm.todo.text)
  })

  test('done', async () => {
    const done = wrapper.find('[data-testid="todo-done"]')
    const todoItem = wrapper.find('[data-testid="todo-item"]')
    expect(done.element.checked).toBeTruthy()
    expect(todoItem.classes()).toContain('completed')

    await done.setChecked(false)
    expect(todoItem.classes('completed')).toBeFalsy()
  })

  test('delete todo', async () => {
    const deleteBtn = wrapper.find('button[data-testid="delete"]')
    await deleteBtn.trigger('click')
    expect(wrapper.emitted()['delete-todo']).toBeTruthy()
    expect(wrapper.emitted()['delete-todo'][0][0]).toBe(wrapper.vm.todo.id)
  })

  test('edit todo style', async () => {
    const label = wrapper.find('label[data-testid="todo-text"]')
    const todoItem = wrapper.find('li[data-testid="todo-item"]')
    const todoEdit = wrapper.find('input[data-testid="todo-edit"]')
    await label.trigger('dblclick')
    expect(todoItem.classes()).toContain('editing')

    await todoEdit.trigger('blur')
    expect(todoItem.classes('editing')).toBeFalsy()
  })

  test('save edit todo', async () => {
    const label = wrapper.find('label[data-testid="todo-text"]')
    const todoEdit = wrapper.find('input[data-testid="todo-edit"]')
    await label.trigger('dblclick')

    // 编辑文本框中的内容展示
    expect(todoEdit.element.value).toBe(wrapper.vm.todo.text)

    // 修改文本框的值
    const text = 'hello'
    await todoEdit.setValue(text)

    // 触发回车保存事件
    await todoEdit.trigger('keyup.enter')

    // 断言数据被修改了
    // expect(wrapper.vm.todo.text).toBe(text)
    expect(wrapper.emitted()['edit-todo']).toBeTruthy()
    expect(wrapper.emitted()['edit-todo'][0][0]).toEqual({
      id: wrapper.vm.todo.id,
      text
    })

    // 确保取消了编辑状态
    expect(wrapper.vm.isEditing).toBeFalsy()
  })

  test('cancel edit todo', async () => {
    const label = wrapper.find('label[data-testid="todo-text"]')
    const todoEdit = wrapper.find('input[data-testid="todo-edit"]')

    await label.trigger('dblclick')
    const text = wrapper.vm.todo.text
    await todoEdit.setValue('bbb')
    // 触发取消的事件
    await todoEdit.trigger('keyup.esc')

    // 验证字段没有被修改
    expect(wrapper.vm.todo.text).toBe(text)
    // 验证编辑状态被取消
    expect(wrapper.vm.isEditing).toBeFalsy()
  })

  test('snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
})
