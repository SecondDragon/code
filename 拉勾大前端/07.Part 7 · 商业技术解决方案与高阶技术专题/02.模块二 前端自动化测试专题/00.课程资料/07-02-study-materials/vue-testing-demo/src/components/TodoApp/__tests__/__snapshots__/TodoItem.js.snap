// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem.vue snapshot 1`] = `
<li data-testid="todo-item" class="completed">
  <div class="view"><input data-testid="todo-done" type="checkbox" class="toggle"> <label data-testid="todo-text">play</label> <button data-testid="delete" class="destroy"></button></div> <input data-testid="todo-edit" class="edit">
</li>
`;
